med-assist-backend/src/test/java/inc/evil/medassist/common/validation/AtLeastOneNotNullConstraintValidatorTest.java

```diff
@@ -0,0 +1,104 @@
+package inc.evil.medassist.common.validation;
+
+import lombok.Data;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.Spy;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import javax.validation.ConstraintValidatorContext;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class AtLeastOneNotNullConstraintValidatorTest {
+    @Spy
+    private AtLeastOneNotNullConstraintValidator atLeastOneNotNullConstraintValidator;
```

Comment:
This test is kind of too tied to the actual implementation. We can try to verify the behavior instead, something like this:
```java

public class AtLeastOneNotNullConstraintValidatorTest {
    private Validator validator;

    @BeforeEach
    private void setUp() {
        ValidatorFactory factory = Validation.buildDefaultValidatorFactory();
        validator = factory.getValidator();
    }
    
    @Test
    public void isValid_whenOneFieldIsNotNull_isTrue() {
        ObjectToValidate object = new ObjectToValidate("field1", null);

        Set<ConstraintViolation<ObjectToValidate>> actualViolations = validator.validate(object);

        assertThat(actualViolations).isEmpty();
    }

    @Test
    public void isValid_whenOtherFieldIsNotNull_isTrue() {
        ObjectToValidate object = new ObjectToValidate(null, "field2");

        Set<ConstraintViolation<ObjectToValidate>> actualViolations = validator.validate(object);

        assertThat(actualViolations).isEmpty();
    }

    @Test
    public void isValid_whenBothFieldsNotNull_isTrue() {
        ObjectToValidate object = new ObjectToValidate("field1", "field2");

        Set<ConstraintViolation<ObjectToValidate>> actualViolations = validator.validate(object);

        assertThat(actualViolations).isEmpty();
    }

    @Test
    public void isValid_whenBothFieldsNull_addsConstraintViolations() {
        ObjectToValidate object = new ObjectToValidate(null, null);

        Set<ConstraintViolation<ObjectToValidate>> violations = validator.validate(object);

        List<Violation> actualViolations = violations.stream()
                .map(v -> new Violation(v.getMessage(), v.getPropertyPath().toString()))
                .toList();
        Violation[] expectedViolations = {
                new Violation("at least should not be null", "field1"),
                new Violation("at least should not be null", "field2")
        };
        assertThat(actualViolations).containsExactlyInAnyOrder(expectedViolations);
    }

    record Violation(String message, String path) {}

    @Data
    @AllArgsConstructor
    @AtLeastOneNotNull(fields = {"field1", "field2"})
    class ObjectToValidate {
        private String field1;
        private String field2;
    }
}

``` 

Way simpler in my opinion

Reply:
It also respects the principle: "Never mock something you don't own" :)

Reply:
agree